src/util-landlock.c

```diff
@@ -0,0 +1,189 @@
+#define _LANDLOCK_ACCESS_FS_READ (LANDLOCK_ACCESS_FS_READ_FILE | LANDLOCK_ACCESS_FS_READ_DIR)
+
+#define _LANDLOCK_SURI_ACCESS_FS_WRITE                                                             \
+    (LANDLOCK_ACCESS_FS_WRITE_FILE | LANDLOCK_ACCESS_FS_MAKE_DIR | LANDLOCK_ACCESS_FS_MAKE_REG |   \
+            LANDLOCK_ACCESS_FS_REMOVE_FILE | LANDLOCK_ACCESS_FS_MAKE_SOCK)
+
+static inline int LandlockCreateRuleset()
+{
+    int ruleset_fd;
+
+    struct landlock_ruleset_attr ruleset_attr = {
+        .handled_access_fs =
+                _LANDLOCK_ACCESS_FS_READ | _LANDLOCK_ACCESS_FS_WRITE | LANDLOCK_ACCESS_FS_EXECUTE,
+    };
+
+    int abi;
+    abi = landlock_create_ruleset(NULL, 0, LANDLOCK_CREATE_RULESET_VERSION);
+    if (abi < 2) {
+        ruleset_attr.handled_access_fs &= ~LANDLOCK_ACCESS_FS_REFER;
+    }
+
+    ruleset_fd = landlock_create_ruleset(&ruleset_attr, sizeof(ruleset_attr), 0);
+    if (ruleset_fd < 0) {
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't create landlock ruleset");
+    }
+    return ruleset_fd;
+}
+
+static inline void LandlockEnforceRuleset(int ruleset_fd)
+{
+    prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0);
+    if (landlock_restrict_self(ruleset_fd, 0))
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't self restrict: %s", strerror(errno));
+}
+
+static int LandlockSandboxingAddRule(int ruleset_fd, const char *directory, uint64_t permission)
+{
+    struct landlock_path_beneath_attr path_beneath = {
+        .allowed_access = permission,
+    };
+
+    int dir_fd = open(directory, O_PATH | O_CLOEXEC | O_DIRECTORY);
+    if (dir_fd != -1)
+        path_beneath.parent_fd = dir_fd;
+    else {
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't open %s", directory);
+        return -1;
+    }
+
+    if (landlock_add_rule(ruleset_fd, LANDLOCK_RULE_PATH_BENEATH, &path_beneath, 0)) {
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't add write rule: %s", strerror(errno));
+        close(dir_fd);
+        return -1;
+    }
+
+    if (dir_fd == -1)
+        close(dir_fd);
+
+    return 0;
+}
+
+static inline void LandlockSandboxingWritePath(int ruleset_fd, const char *directory)
+{
+    if (LandlockSandboxingAddRule(ruleset_fd, directory, _LANDLOCK_SURI_ACCESS_FS_WRITE) == 0) {
+        SCLogInfo("Added write permission to '%s'", directory);
+    }
+}
+
+static inline void LandlockSandboxingReadPath(int ruleset_fd, const char *directory)
+{
+    if (LandlockSandboxingAddRule(ruleset_fd, directory, _LANDLOCK_ACCESS_FS_READ) == 0) {
+        SCLogInfo("Added read permission to '%s'", directory);
+    }
+}
+
+void LandlockSandboxing(SCInstance *suri)
+{
+    /* Read configuration variable and exit if no enforcement */
+    int conf_status;
+    ConfGetBool("landlock.enabled", &conf_status);
+    if (!conf_status) {
+        SCLogNotice("Landlock is not enabled in configuration");
+        return;
+    }
+    int ruleset_fd = LandlockCreateRuleset();
+
+    LandlockSandboxingWritePath(ruleset_fd, ConfigGetLogDirectory());
```

Comment:
If we are implicitly adding the log directory, do we need to include it in `landlock.directories.write`?

Reply:
Yes, this is not needed. Good catch!